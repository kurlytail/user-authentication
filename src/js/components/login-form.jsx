import React, { Component } from 'react';
import PropTypes from 'prop-types';

class LoginForm extends Component {
    handleChange(event) {
        this.setState({ [event.target.id.replace('auth-signin-', '')]: event.target.value });
    }

    render() {
        return (
            <form className="auth-form" id="auth-signin-form" encType="utf8">
                <div className="form-group">
                    <div className="form-label-group">
                        <input
                            type="text"
                            id="auth-signin-email"
                            className="form-control"
                            placeholder="Username"
                            required
                            autoFocus
                            onChange={event => this.handleChange(event)}
                        />
                        <label htmlFor="auth-signin-email">Username</label>
                    </div>
                </div>

                <div className="form-group">
                    <div className="form-label-group">
                        <input
                            type="password"
                            id="auth-signin-password"
                            className="form-control"
                            placeholder="Password"
                            required
                            onChange={event => this.handleChange(event)}
                        />
                        <label htmlFor="auth-signin-password">Password</label>
                    </div>
                </div>

                <div className="form-group">
                    <button
                        className="btn btn-lg btn-primary btn-block"
                        id="auth-signin-button"
                        onClick={e => {
                            this.props.login(this.state.email, this.state.password);
                            e.preventDefault();
                        }}
                    >
                        Sign In
                    </button>
                </div>

                <div className="form-group text-center">
                    <div className="custom-control custom-control-inline custom-checkbox">
                        <input type="checkbox" className="custom-control-input" id="remember-me" />
                        <label className="custom-control-label" htmlFor="remember-me">
                            Keep me signed in
                        </label>
                    </div>
                </div>

                <div className="text-center pt-3">
                    <a className="link">Forgot Username?</a>
                    <span className="mx-2">Â·</span>
                    <a className="link">Forgot Password?</a>
                </div>
            </form>
        );
    }
}

LoginForm.propTypes = {
    login: PropTypes.func,
    pending: PropTypes.any
};

export default LoginForm;
