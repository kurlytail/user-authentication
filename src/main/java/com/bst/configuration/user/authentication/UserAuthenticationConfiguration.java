package com.bst.configuration.user.authentication;

import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;

@Configuration
@ComponentScan("com.bst.user")
public class UserAuthenticationConfiguration {

}
