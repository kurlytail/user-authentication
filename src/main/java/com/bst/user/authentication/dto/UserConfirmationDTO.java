
package com.bst.user.authentication.dto;

import javax.validation.constraints.Email;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

public class UserConfirmationDTO {

	@Email
	@NotNull
	@NotEmpty
	private String email;

	@NotNull
	@NotEmpty
	private String password;

	public String getPassword() {
		return password;
	}

	public void setPassword(String password) {
		this.password = password;
	}

	public String getEmail() {
		return email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

}
