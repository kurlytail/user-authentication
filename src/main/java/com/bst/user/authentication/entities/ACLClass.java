package com.bst.user.authentication.entities;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import javax.validation.constraints.NotNull;

@Entity
@Table(name = "acl_class", uniqueConstraints = { @UniqueConstraint(columnNames = { "class" }) })
public class ACLClass {
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	@NotNull
	private Long id;

	@Column(name = "class")
	private String aclClass;
}
