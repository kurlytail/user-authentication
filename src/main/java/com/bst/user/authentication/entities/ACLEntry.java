package com.bst.user.authentication.entities;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import javax.validation.constraints.NotNull;

@Entity
@Table(name = "acl_class", uniqueConstraints = {
		@UniqueConstraint(columnNames = { "acl_object_identity", "ace_order" }) })
public class ACLEntry {
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	@NotNull
	private Long id;

	@Column(name = "acl_object_identity")
	private ACLObjectIdentity aclObjectIdentity;

	@Column(name = "object_id_identity")
	private Long objectIdIdentity;

	@Column(name = "ace_order")
	private Long aceOrder;

	@Column(name = "sid")
	private ACLSID sid;

	@Column(name = "mask")
	private Long mask;
	
	@Column(name = "granting")
	private Integer granting;
	
	@Column(name = "audit_success")
	private Integer auditSuccess;
	
	@Column(name = "sid")
	private Integer audit_failure;
}
