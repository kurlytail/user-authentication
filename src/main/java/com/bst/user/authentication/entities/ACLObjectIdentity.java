package com.bst.user.authentication.entities;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import javax.validation.constraints.NotNull;

@Entity
@Table(name = "acl_object_identity", uniqueConstraints = {
		@UniqueConstraint(columnNames = { "object_id_class", "object_id_identity" }) })
public class ACLObjectIdentity {

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	@NotNull
	private Long id;

	@Column(name = "object_id_class")
	private ACLClass aclClass;

	@Column(name = "parent_object")
	private ACLObjectIdentity parentObject;

	@Column(name = "owner_sid")
	private ACLSID ownerSID;

	@Column(name = "entries_inheriting")
	private Integer entriesInheriting;
	
	@Column (name = "object_id_identity")
	private Long objectIdIdentity;

}
