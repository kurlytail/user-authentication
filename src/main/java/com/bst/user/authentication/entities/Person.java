package com.bst.user.authentication.entities;

import java.io.Serializable;

import javax.persistence.Column;
import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;

@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "userType")
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
public class Person implements Serializable {

	/**
	 *
	 */
	private static final long serialVersionUID = 1L;

	@Column(unique = true)
	private String email;

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	private String name;

	private String password;

	public Person() {
	}

	public Person(final String email) {
		this.email = email;
	}

	@Override
	public boolean equals(final Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (this.getClass() != obj.getClass()) {
			return false;
		}
		final Person other = (Person) obj;
		if (this.email == null) {
			if (other.email != null) {
				return false;
			}
		} else if (!this.email.equals(other.email)) {
			return false;
		}
		if (this.id == null) {
			if (other.id != null) {
				return false;
			}
		} else if (!this.id.equals(other.id)) {
			return false;
		}
		if (this.name == null) {
			if (other.name != null) {
				return false;
			}
		} else if (!this.name.equals(other.name)) {
			return false;
		}
		if (this.password == null) {
			if (other.password != null) {
				return false;
			}
		} else if (!this.password.equals(other.password)) {
			return false;
		}
		return true;
	}

	public String getEmail() {
		return this.email;
	}

	public Long getId() {
		return this.id;
	}

	public String getName() {
		return this.name;
	}

	public String getPassword() {
		return this.password;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = (prime * result) + ((this.email == null) ? 0 : this.email.hashCode());
		result = (prime * result) + ((this.id == null) ? 0 : this.id.hashCode());
		result = (prime * result) + ((this.name == null) ? 0 : this.name.hashCode());
		result = (prime * result) + ((this.password == null) ? 0 : this.password.hashCode());
		return result;
	}

	public void setName(final String name) {
		this.name = name;
	}

	public void setPassword(final String password) {
		this.password = password;
	}

	@Override
	public String toString() {
		return "Person [id=" + this.id + ", email=" + this.email + ", name=" + this.name + "]";
	}

}
