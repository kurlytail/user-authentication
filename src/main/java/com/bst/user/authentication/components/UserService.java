package com.bst.user.authentication.components;

import javax.transaction.Transactional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;

import com.bst.user.authentication.entities.LocalUserDetails;
import com.bst.user.authentication.entities.Person;
import com.bst.user.authentication.repositories.PersonRepository;

@Component
public class UserService implements UserDetailsService {
	
	@Autowired
	private PersonRepository userRepository;
	
	@Autowired
	private PasswordEncoder passwordEncoder;

	@Override
	@Transactional
	public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
		
		Person user = userRepository.findByEmail(username);
		if (user != null) {
			return new LocalUserDetails(user);
		}
		throw new UsernameNotFoundException("Bad credentials");
	}
	
	@Transactional
	public Person loadUser(String username) throws UsernameNotFoundException {
		
		Person user = userRepository.findByEmail(username);
		if (user != null) {
			return user;
		}
		throw new UsernameNotFoundException("Bad credentials");
	}
	
	@Transactional
	public Person createUser(String email, String name, String password) {
		Person user = new Person(email);
		user.setName(name);
		user.setPassword(passwordEncoder.encode(password));
		
		userRepository.save(user);
		return user;
	}
}
