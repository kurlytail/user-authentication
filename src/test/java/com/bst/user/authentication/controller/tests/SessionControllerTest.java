package com.bst.user.authentication.controller.tests;

import java.util.List;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.TestExecutionListeners;
import org.springframework.test.context.TestExecutionListeners.MergeMode;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import com.bst.configuration.user.authentication.UserAuthenticationConfiguration;
import com.bst.user.authentication.components.UserService;
import com.bst.user.authentication.dto.UserConfirmationDTO;
import com.bst.user.authentication.entities.Person;
import com.bst.utility.testlib.SnapshotListener;

@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = { UserAuthenticationConfiguration.class, WebSecurityConfiguration.class })
@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
@TestExecutionListeners(listeners = SnapshotListener.class, mergeMode = MergeMode.MERGE_WITH_DEFAULTS)
@TestPropertySource("classpath:session-controller-test.properties")
public class SessionControllerTest {

	private HttpHeaders requestHeaders;

	@Autowired
	private TestRestTemplate testRestTemplate;

	@Autowired
	private UserService userService;

	@Test
	public void sessionStateMachineCheck() throws Exception {

		final UserConfirmationDTO dto = new UserConfirmationDTO();
		dto.setEmail("john@doe.com");
		dto.setPassword("password");

		this.userService.createUser(dto.getEmail(), "John Doe", dto.getPassword());

		ResponseEntity<Person> getResponse = this.testRestTemplate.getForEntity("/auth/session", Person.class);
		this.useSessionId(getResponse.getHeaders());
		SnapshotListener.expect(getResponse.getStatusCode()).toMatchSnapshot();
		SnapshotListener.expect(getResponse.getBody()).toMatchSnapshot();

		final ResponseEntity<Person> postResponse = this.testRestTemplate.exchange("/auth/session", HttpMethod.POST,
				new HttpEntity<>(dto, this.requestHeaders), Person.class);
		SnapshotListener.expect(postResponse.getStatusCode()).toMatchSnapshot();
		SnapshotListener.expect(postResponse.getBody()).toMatchSnapshot();

		getResponse = this.testRestTemplate.exchange("/auth/session", HttpMethod.GET,
				new HttpEntity<>(this.requestHeaders), Person.class);
		SnapshotListener.expect(getResponse.getStatusCode()).toMatchSnapshot();
		SnapshotListener.expect(getResponse.getBody()).toMatchSnapshot();

		ResponseEntity<Person> deleteResponse = this.testRestTemplate.exchange("/auth/session", HttpMethod.DELETE,
				new HttpEntity<>(dto, this.requestHeaders), Person.class);
		SnapshotListener.expect(deleteResponse.getStatusCode()).toMatchSnapshot();
		SnapshotListener.expect(deleteResponse.getBody()).toMatchSnapshot();

		getResponse = this.testRestTemplate.exchange("/auth/session", HttpMethod.GET,
				new HttpEntity<>(this.requestHeaders), Person.class);
		SnapshotListener.expect(getResponse.getStatusCode()).toMatchSnapshot();
		SnapshotListener.expect(getResponse.getBody()).toMatchSnapshot();

		deleteResponse = this.testRestTemplate.exchange("/auth/session", HttpMethod.DELETE,
				new HttpEntity<>(dto, this.requestHeaders), Person.class);
		SnapshotListener.expect(deleteResponse.getStatusCode()).toMatchSnapshot();
		SnapshotListener.expect(deleteResponse.getBody()).toMatchSnapshot();
	}

	private void useSessionId(final HttpHeaders headers) {
		List<String> cookies = headers.get("Cookie");
		if (cookies == null) {
			cookies = headers.get("Set-Cookie");
		}
		final String cookie = cookies.get(cookies.size() - 1);
		final int start = cookie.indexOf('=');
		final int end = cookie.indexOf(';');

		final String sessionId = cookie.substring(start + 1, end);
		this.requestHeaders = new HttpHeaders();
		this.requestHeaders.add("Cookie", "JSESSIONID=" + sessionId);
	}

}
