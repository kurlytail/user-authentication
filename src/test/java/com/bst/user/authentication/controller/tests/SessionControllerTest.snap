{"sessionStateMachineCheck":["\"NOT_FOUND\"","{\"id\":null,\"email\":null,\"name\":null,\"password\":null}","\"OK\"","{\"id\":1,\"email\":\"john@doe.com\",\"name\":\"John Doe\",\"password\":null}","\"OK\"","{\"id\":1,\"email\":\"john@doe.com\",\"name\":\"John Doe\",\"password\":null}","\"OK\"","null","\"NOT_FOUND\"","{\"id\":null,\"email\":null,\"name\":null,\"password\":null}","\"NOT_FOUND\"","{\"id\":null,\"email\":null,\"name\":null,\"password\":null}"]}